infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/Bootstrapper.java

```diff
@@ -0,0 +1,176 @@
+        this.eventService = eventService;
+        this.machineName = machineName;
+        this.runtimeIdentity = runtimeIdentity;
+        this.dockerMachine = dockerMachine;
+        this.bootstraplatch = new CountDownLatch(1);
+        this.endpointId = new AtomicInteger();
+        this.wsIntallerEndpoint = apiEndpoint.replace("http", "ws").replace("wsmaster/api", "wsmaster")
+                                  + "/installer/websocket/";
+        this.bootstrapperStatusListener = event -> {
+            BootstrapperStatus status = event.getStatus();
+
+            //skip starting status event
+            if (status.equals(BootstrapperStatus.DONE) || status.equals(BootstrapperStatus.FAILED)) {
+                //check boostrapper belongs to current runtime and machine
+                RuntimeIdentityDto runtimeId = event.getRuntimeId();
+                if (event.getMachineName().equals(machineName)
+                    && runtimeIdentity.getEnvName().equals(runtimeId.getEnvName())
+                    && runtimeIdentity.getOwner().equals(runtimeId.getOwner())
+                    && runtimeIdentity.getWorkspaceId().equals(runtimeId.getWorkspaceId())) {
+
+                    resultEvent = event;
+                    bootstraplatch.countDown();
+                }
+            }
+        };
+    }
+
+    /**
+     * Bootstraps agents installers and wait while they finished.
+     *
+     * @param agents
+     *         agents to install
+     * @throws InfrastructureException
+     *         when bootstrapping timeout reached
+     * @throws InfrastructureException
+     *         when bootstrapping failed
+     * @throws InfrastructureException
+     *         when any other error occurs while bootstrapping
+     */
+    public void bootstrap(List<AgentImpl> agents) throws InfrastructureException {
+        injectBootstrapper(agents);
+
+        this.eventService.subscribe(bootstrapperStatusListener, BootstrapperStatusEvent.class);
+        boolean bootstrappingFinished;
+        try {
+            String endpoint = wsIntallerEndpoint + endpointId.getAndIncrement();
+            dockerMachine.exec(BOOTSTRAPPER_BASE_DIR + BOOTSTRAPPER_FILE +
+                               " -machine-name " + machineName +
+                               " -runtime-id " + String.format("%s:%s:%s", runtimeIdentity.getWorkspaceId(),
+                                                               runtimeIdentity.getEnvName(),
+                                                               runtimeIdentity.getOwner()) +
+                               " -push-endpoint " + endpoint +
+                               " -push-logs-endpoint " + endpoint +
+                               " -file " + BOOTSTRAPPER_BASE_DIR + CONFIG_FILE,
+                               null);
+
+            //waiting for DONE or FAILED bootstrapper status event
+            bootstrappingFinished = bootstraplatch.await(3, TimeUnit.MINUTES);
```

Comment:
Consider using `CompletableFuture<BootstrapperStatusEvent>` instead of latch + volatile `resultEvent`.
The approach would be:
```java
Bootstrapper( ... ) {
    /// ...
   this.finishedFuture = new CompletableFuture<>();
}
```

When `DONE` | `FAILED` event received, completion propagated, e.g.
```java
private class FinishedEventSubscriber implements EventSubscriber<BootstrapperStatusEvent> {
    @Override
    public void onEvent(BootstrapperStatusEvent event) {
        BootstrapperStatus status = event.getStatus();
        //skip starting status event
        if (status.equals(BootstrapperStatus.DONE) || status.equals(BootstrapperStatus.FAILED)) {
           //check boostrapper belongs to current runtime and machine
            RuntimeIdentityDto runtimeId = event.getRuntimeId();
            if (event.getMachineName().equals(machineName)
                && runtimeIdentity.getEnvName().equals(runtimeId.getEnvName())
                && runtimeIdentity.getOwner().equals(runtimeId.getOwner())
                && runtimeIdentity.getWorkspaceId().equals(runtimeId.getWorkspaceId())) {

                finishedFuture.complete(bootstrapperEvent); // <- future completion
                eventService.unsubscribe(this); // <- self unsubscribe
            }
        }
    }
}
```
and then instead of calling `latch.await(10, TimeUnit.MINUTES)` call
```java
BootstrapperStatusEvent event = finishedFuture.get(10, TimeUnit.MINUTES);
```
This approach simplifies safe result event publication from event subscriber to its client.